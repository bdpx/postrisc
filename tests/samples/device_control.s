
.text
    write	"test control device memory-mapped registers"
    alloc	96

    ; device_control base address
    ldi.l	%r24, DEVICE_CONFIG_VIRT_BASE

    write	"test pci"

    ldi.l	%r21, 0x1234567890abcdef

    ldz.d	%r20, %r24, DEVICE_CONTROL_DID
    write	"mem[DEVICE_CONTROL_DID] %x64(r20)"
    st.d	%r21, %r24, DEVICE_CONTROL_DID
    ldz.d	%r20, %r24, DEVICE_CONTROL_DID
    write	"mem[DEVICE_CONTROL_DID] %x64(r20)"

    ldz.d	%r20, %r24, DEVICE_CONTROL_CMD
    write	"mem[DEVICE_CONTROL_CMD] %x64(r20)"
    st.d	%r21, %r24, DEVICE_CONTROL_CMD
    ldz.d	%r20, %r24, DEVICE_CONTROL_CMD
    write	"mem[DEVICE_CONTROL_CMD] %x64(r20)"

    ldz.d	%r20, %r24, DEVICE_CONTROL_ARRAY_ADDRESS
    write	"mem[DEVICE_CONTROL_ARRAY_ADDRESS] (r20)"

    ldz.d	%r20, %r24, DEVICE_CONTROL_ARRAY_LEN
    write	"mem[DEVICE_CONTROL_ARRAY_LEN] %i64(r20)"

    ldi	%r22, '\n'

    write	"test command"
    ldi.l	%r21, 0xabcdef1234567890
    st.d	%r21, %r24, DEVICE_CONTROL_CMD

    write	"end_device_control_test"
.end
