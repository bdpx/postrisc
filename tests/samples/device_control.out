test control device memory-mapped registers
test pci
read pci_did
mem[DEVICE_CONTROL_DID] 0000000000000000
writing pci_did: 0x1234567890abcdef
read pci_did
mem[DEVICE_CONTROL_DID] 1234567890abcdef
read pci_cmd
mem[DEVICE_CONTROL_CMD] 0000000000000000
writing pci_cmd: 0x1234567890abcdef
read pci_cmd
mem[DEVICE_CONTROL_CMD] 1234567890abcdef
read device_array_address
mem[DEVICE_CONTROL_ARRAY_ADDRESS] (r20)
read device_array_len
mem[DEVICE_CONTROL_ARRAY_LEN] 24576
test command
writing pci_cmd: 0xabcdef1234567890
end_device_control_test
